### DISCUSSION TITLE
Import parallels macOS vms

### DISCUSSION THREAD

**Discussion by matthewruzzi** [utmapp/UTM#4113] [Ideas]
Parallels and UTM seem to use very similar formats for macOS vms. Would it be possible to allow importing parallels vms into UTM?

![Screen Shot 2022-06-09 at 12 35 50 PM](https://user-images.githubusercontent.com/68619790/172930371-551684be-a28b-4f6a-b9d3-ab851d8fecd6.png)
![Screen Shot 2022-06-09 at 12 35 29 PM](https://user-images.githubusercontent.com/68619790/172930385-eec34f6b-b150-4a31-8721-a1beb66d9fc3.png)

**Comment by xlkuet** [comment]
I have the same question. I actually want to shift from my parallel PVM to UTM. However, not found any solution till now. Looking forward the UTM community to help in this regard.

**Comment by tarasis** [comment]
Something I would love to see as well. I'd like to test Windows 11 arm support without having to create a whole new VM.

**Comment by hcpcrose** [comment]
I believe they are very similar.  At least for any ARM based VM running on apple silicon Macs.  Apple now includes a very robust virtualization framework.  If you are comfortable with the Mac Terminal you could create and run VM's without an app like Parallels or UTM.  I prefer a nice GUI, which is why I am willing to buy and support the UTM app.  

Here's a nice article from Ars Technica that explains it in a little more depth:  https://arstechnica.com/gadgets/2022/07/how-to-use-free-virtualization-apps-to-safely-test-the-macos-ventura-betas

Here's the Apple documentation:  https://developer.apple.com/documentation/virtualization

**Comment by osy** [comment]
The format doesn’t matter. The difference is the hardware they emulate and it’s not compatible. It’s like if you take your hard drive out of one computer and put it into another. It may work but likely there’ll be lots of driver issues. Windows (older versions) doesn’t like swapping hardware.

**Reply by matthewruzzi** [reply]
Both Parallels, UTM, and other options like [KhaosT/MacVM](https://github.com/KhaosT/MacVM) and [VirtualBuddy](https://github.com/insidegui/VirtualBuddy) all use the apple virtualization framework for macOS guest vms, so it seems like they would all emulate the same hardware.

**Reply by hcpcrose** [reply]
Yes.  That is also true for VMware Fusion, which uses apple virtualization.  

But UTM is built on top of QEMU.  Apple's virtualization framework is pretty new, but QEMU has been around for a long, long time. QEMU has its own virtualization. Beyond that you can also emulate non native operating systems with UTM/QEMU.  

That's why when you create a new UTM VM from scratch the first dialogue asks if you want to "virtualize" or "emulate".  If you choose virtualize, I believe UTM still defaults to using QEMU virtualization.  

So, for now, the UTM developers have their hands full. Eventually, when Apple's virtualization platform matures, and it can become the default, things should simplify for virtualized VMs. But for now virtualized UTM VM's still contain different virtualized hardware than VM's created with apple virtualization.

**Reply by agenoras** [reply]
Actually, swapping a drive with Windows into another computer is not as difficult as you make it out to be. You run "sysprep" before and you are done. So, it is probably just not a good analogy.

**Comment by hcpcrose** [comment]
This discussion started out as an enhancement request to import Parallels VM's.  That would be the ultimate solution to this issues.  I don't want the main contributor (osy) to get distracted from his work, so I'll try to keep this short.  🙂

Until this enhancement is complete, some of us may want to work on moving our machines manually.  osy makes a good point about the differences in emulated hardware between various VM platforms.  But it doesn't mean VM's aren't portable across platforms, they just may not be *easily* portable.  😂  I've worked with every virtualization platform that exists, so I'm pretty comfortable hacking them.  YMMV.  

I have had success in moving Windows and ARM Mac virtual disk files back and forth from Vmware Fusion to Parallels on Apple Silicon, without having to reinstall from scratch.  It takes a little work.  Important to uninstall the current guest tools first, then reinstall the new tools after migration.  This works well for Windows or macOS, haven't tried yet with Linux.  

I was thinking the same might be true for moving native ARM Windows or Mac VMs from Parallels or Fusion to UTM.  (Eventually I plan to experiment with this, but haven't yet.  I would like to eventually move to UTM, and off Parallels).    

Of course portability to and from Parallels or Fusion is only possible for ARM based OS's running natively on Apple silicon.  

Running emulated OS's (like Solaris SPARC or Mac 9.2 PPC) can only be run using UTM (because UTM uses QEMU).  

I hope osy eventually has time to work on a UTM import capability.  UTM is a great app, and that would make it even better.

**Reply by osy** [reply]
I think if it’s limited to Windows ARM or just arm machines it may be very easy. However, I’m thinking about also x86 and moving from parallels virtualization to QEMU emulation and so on.

**Reply by hcpcrose** [reply]
Whoa!  That would be very ambitious.  And awesome!  Maybe start with ARM, and add the others later... (just a suggestion)   Good luck, and keep up the good work.

**Comment by ballo** [comment]
I attempted to convert my (ARM64) Linux install from Parallels into UTM several times with no success. UTM refuses to boot anything. Has anyone managed to do a successful conversion?

**Comment by epheterson** [comment]
I got it to work, you have to be sure to delete snapshots for the conversion to be able to run.

ChatGPT and I made a script.

Filename: convert_to_qcow2.sh

Usage: 
`brew install qemu`
`chmod u+x convert_to_qcow2.sh`
`./convert_to_qcow2.sh <path_to_pvm_file>`

```
#!/bin/bash

# Ensure the script is run with a Parallels VM file argument
if [ -z "$1" ]; then
  echo "Usage: $0 <path_to_parallels_pvm>"
  exit 1
fi

PVM_PATH=$1

# Define paths and filenames
PRL_DISK_TOOL="/Applications/Parallels Desktop.app/Contents/MacOS/prl_disk_tool"
QCOW2_FILE="${PVM_PATH%.pvm}.qcow2"

# Ensure prlctl is available
if ! command -v prlctl &> /dev/null; then
  echo "prlctl command not found. Please ensure Parallels Tools are installed and in your PATH."
  exit 1
fi

# Delete all snapshots
echo "Deleting snapshots..."
prlctl snapshot-delete "$PVM_PATH" --id all

# Find the HDD file within the PVM package
HDD_FILE=$(find "$PVM_PATH" -name '*.hdd' -print -quit)

# Ensure the HDD file was found
if [ -z "$HDD_FILE" ]; then
  echo "No HDD file found in the specified Parallels VM package."
  exit 1
fi

# Convert HDD to plain format
echo "Converting HDD to plain format..."
"$PRL_DISK_TOOL" convert --hdd "$HDD_FILE" --plain

# Find the HDS file from the converted HDD package
PLAIN_HDS_FILE=$(find "$HDD_FILE" -name '*.hds' -print -quit)

# Ensure the HDS file was found
if [ -z "$PLAIN_HDS_FILE" ]; then
  echo "No HDS file found in the converted HDD package."
  exit 1
fi

# Convert HDS to QCOW2 format using qemu-img
echo "Converting HDS to QCOW2 format..."
qemu-img convert -O qcow2 "$PLAIN_HDS_FILE" "$QCOW2_FILE"

echo "Conversion complete. QCOW2 file is located at: $QCOW2_FILE"
```
